beacon-chain/rpc/beacon_server.go

```diff
@@ -210,18 +210,81 @@ func (bs *BeaconServer) PendingDeposits(ctx context.Context, _ *ptypes.Empty) (*
 	}
 	// Only request deposits that have passed the ETH1 follow distance window.
 	bNum = bNum.Sub(bNum, big.NewInt(int64(params.BeaconConfig().Eth1FollowDistance)))
-	return &pb.PendingDepositsResponse{PendingDeposits: bs.beaconDB.PendingDeposits(ctx, bNum)}, nil
+	pendingDeps := bs.beaconDB.PendingDeposits(ctx, bNum)
+	allDeps := bs.beaconDB.AllDeposits(ctx, bNum)
+	if len(allDeps) == 0 {
+		return &pb.PendingDepositsResponse{PendingDeposits: nil}, nil
+	}
+
+	// Need to fetch if the deposits up to the state's latest eth 1 data matches
+	// the number of all deposits in this RPC call. If not, then we return nil.
+	beaconState, err := bs.beaconDB.State(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch beacon state: %v", err)
+	}
+	h := bytesutil.ToBytes32(beaconState.LatestEth1Data.BlockHash32)
+	_, latestEth1DataHeight, err := bs.powChainService.BlockExists(ctx, h)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch eth1data height: %v", err)
+	}
+	upToLatestEth1DataDeposits := bs.beaconDB.AllDeposits(ctx, latestEth1DataHeight)
+	if len(upToLatestEth1DataDeposits) != len(allDeps) {
```

Comment:
I don't understand this condition. Why do you do this?

Reply:
If the state's latest eth1 data's block hash has a height of 100, we fetch all the deposits up to height 100. If this doesn't match the number of deposits stored in the cache, the generated trie will not be the same and root will fail to verify. This can happen in a scenario where we perhaps have a deposit from height 101, so we want to avoid any possible mismatches in these lengths.

Reply:
Nice explanation, we should add a comment for it : )